### DISCUSSION TITLE
Manually locate libffi for pyenv installation

### DISCUSSION THREAD

**Discussion by mrlasdt** [pyenv/pyenv#2719] [Q&A]
I am a beginner in pyenv. Every time I install a new Python version, I get the following output
```
>> from _ctypes import Union, Structure, Array
ModuleNotFoundError: No module named '_ctypes'
WARNING: The Python ctypes extension was not compiled. Missing the libffi lib?
```
Based on my search, the error is because the libffi-dev library is not installed, and a simple reinstallation should work. But, I don't have the root permission to install it.
```
$ dpkg -L libffi            
dpkg-query: package 'libffi' is not installed
```
Then, I tried to find a manual approach. In the built-in Python of the system, I see that the ctypes package can be imported
```
$ /usr/bin/python -c "import ctypes.util; print(ctypes.util.find_library('ffi'))"
libffi.so.7
```
I tried to find the libffi.so.7 file and get
```
$ find /usr/lib -name "libffi.so*"
/usr/lib/x86_64-linux-gnu/libffi.so.7.1.0
/usr/lib/x86_64-linux-gnu/libffi.so.7
```
Is there any possible way to proceed further?

**Comment by native-api** [comment]
Since `-dev` is headers and lib files, you can unpack the apt package manually and [pass the path to CPython's Configure and/or CPPFLAGS/LDFLAGS](https://github.com/pyenv/pyenv/tree/master/plugins/python-build#special-environment-variables).

https://askubuntu.com/questions/193695/installing-packages-into-local-directory also gives an option to install with Apt into a chroot jail -- which has the same effect but might be more convenient.
